### DISCUSSION TITLE
unifying the standard library

### DISCUSSION THREAD

**Discussion by ellipse12** [c3lang/c3c#1587] [Ideas]
At the moment the way to initialize the data structures in the standard library are not uniform (init_new, new_init, init, etc.). This makes it more difficult to learn the stdlib since you have to memorize all of the different variations for each module. I would like to hear other peoples thoughts on this (and of course @lerno) and ideas as to how it might be unified. The idea I had was something like: new, init, tnew, tinit (in this case new means to allocate the data, and init would initialize it to defaults or values supplied) or other variations of the names (such as create, alloc, etc. for allocation). But let me know what you guys think!

**Comment by lerno** [comment]
It should be new_init and temp_init if it allocates, init otherwise. I don't think there is any init_new anymore (there might be some use of that still in the docs that isn't updated)

But the whole thing is problematic. `new` in `new_init` implies allocations that default to the heap allocator. `temp_init` implies that it is using the temp allocator. Naming for the standard library is hard.

**Reply by ellipse12** [reply]
Oops I was basing it off of the website and v0.6.3. But it would probably still be a good thing to discuss standardizing naming and functions for other things as well (like using interfaces more). It doesn't mean that they have to be implement right now just as it gets closer to 1.0

**Comment by ellipse12** [comment]
Perhaps this would also be a good place to list out what all is needed in the stdlib for 1.0? I can also move this to an issue if that would be better
